MIGRATION_GUIDE.md

```diff
@@ -0,0 +1,778 @@
+This enables typical use cases like dispatching additional initial actions, enabling a UI extension at diagram startup etc.
+The previous approach of overriding the application integration specific diagram widget to implement those use cases is now discouraged.
+Startup logic should be migrated to `IDiagramStartup` implementations which can then be bound in your custom DI module(s).
+For instance, a simple implementation to enable a UI extension on startup could look like this:
+
+```ts
+@injectable()
+export class MyDiagramStartup implements IDiagramStartup {
+    @inject(GLSPActionDispatcher)
+    protected actionDispatcher: GLSPActionDispatcher;
+
+    postModelInitialization(): MaybePromise<void> {
+        this.actionDispatcher.dispatch(SetUIExtensionVisibilityAction.create({ extensionId: 'my-extension', visible: true }));
+    }
+}
+
+
+export const myDiagramModule= new ContainerModule(bind => {
+    bind(TYPES.IDiagramStartup).to(MyDiagramStartup).inSingletonScope();
+}
+```
+
+### Event Listener
+
+With 2.x a generic Event Emitter API has been introduced.
+This API is similar to the event API of Theia/VS Code and allows declaring arbitrary events and corresponding Emitters.
+Interesting parties can subscribe to the event to get notified when the event is fired.
+The existing listener implementations for `SelectionService`, `GLSPCommandStack` & `EditorContextService` have been refactored to use the new API.
+Explicit methods for de-registering listeners are no longer available.
+Instead adopters have to keep track of the disposable that was received when registering the listener and dispose it when the listener is no longer needed.
+
+<details>
+  <summary>1.0.0</summary>
+
+```ts
+@injectable()
+export class MyService {
+    @inject(SelectionService)
+    protected selectionService: SelectionService;
+
+    protected listener: SelectionListener;
+
+    @postConstruct()
+    protected init(): void {
+        this.listener = {
+            selectionChanged: () => {
+                console.log('selection changed');
+            }
+        };
+        this.selectionService.register({
+            selectionChanged: () => {
+                console.log('selection changed');
+            }
+        });
+    }
+
+    dispose(): void {
+        this.selectionService.deregister(this.listener);
+    }
+}
+```
+
+</details>
+<details open>
+  <summary>2.x</summary>
+
+```ts
+@injectable()
+export class MyService implements Disposable {
+    @inject(SelectionService)
+    protected selectionService: SelectionService;
+
+    protected selectionListenerDisposable: Disposable;
+
+    @postConstruct()
+    protected init(): void {
+        this.selectionListenerDisposable = this.selectionService.onSelectionChanged(() => {
+            console.log('selection changed');
+        });
+    }
+
+    dispose(): void {
+        this.selectionListenerDisposable.dispose();
+    }
+}
+```
+
+<details>
+  <summary>List of changes</summary>
```

Comment:
This list is very hidden because they are all collapsed by default. It might be good to add one sentence above to ensure a bit of separation and explanation. For instance, the registration of those listeners (TYPES.ISelectionListener, TYPES.IEditModeListener, TYPES.MouseListener, ...) are now happening as part of the module and not necessarily as part of the initialization.

Of course there is a bit of a risk to introduce circular dependencies, so we need to be careful with that.

Also consider code quoting.

Reply:
>For instance, the registration of those listeners (TYPES.ISelectionListener, TYPES.IEditModeListener, TYPES.MouseListener, ...) are now happening as part of the module and not necessarily as part of the initialization.'

Could you eloborate on that. I don't follow. IMO nothing did change in terms of listener registration between 1.x and 2.x

Reply:
Oh, I may mix something up, I thought we added the bindings for `TYPES.ISelectionListener` in the module with v2 but they might already have been present.